vm/systemSmartContracts/delegation.go

```diff
@@ -1621,9 +1650,51 @@ func (d *delegation) withdraw(args *vmcommon.ContractCallInput) vmcommon.ReturnC
 		return vmcommon.UserError
 	}
 
+	err = d.deleteDelegatorIfNeeded(args.CallerAddr, delegator)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
 	return vmcommon.Ok
 }
 
+func (d *delegation) deleteDelegatorIfNeeded(address []byte, delegator *DelegatorData) error {
+	if d.isOwner(address) {
+		return nil
+	}
+
+	isDelegatorWithoutFunds := len(delegator.ActiveFund) == 0 && len(delegator.UnStakedFunds) == 0
+	if !isDelegatorWithoutFunds {
+		return nil
+	}
+
+	err := d.computeAndUpdateRewards(address, delegator)
+	if err != nil {
+		return err
+	}
+
+	if delegator.UnClaimedRewards.Cmp(zero) == 0 {
+		d.eei.SetStorage(address, nil)
+
+		dStatus, errGet := d.getDelegationStatus()
+		if errGet != nil {
+			return errGet
+		}
+
+		if dStatus.NumUsers > 0 {
+			dStatus.NumUsers--
```

Comment:
This is imported but not tested. Please add an unit test for this (or change an existing one)

Reply:
It is tested. I added tests for this.

Reply:
not tested